File: src/main/java/com/example/retailstore/strategy/AffiliateDiscountStrategy.java
package com.example.retailstore.strategy;

import com.example.retailstore.model.Bill;
import org.springframework.stereotype.Component;

@Component
public class AffiliateDiscountStrategy implements DiscountStrategy {
    private static final double AFFILIATE_DISCOUNT = 0.1;

    @Override
    public double calculateNetPayableAmount(Bill bill) {
        double totalAmount = bill.getTotalAmount();
        double netPayableAmount = totalAmount;

        netPayableAmount -= totalAmount * AFFILIATE_DISCOUNT;

        double discountAmount = (int) (totalAmount / 100) * DISCOUNT_PER_100;
        netPayableAmount -= discountAmount;

        return netPayableAmount;
    }
}

File: src/main/java/com/example/retailstore/strategy/AffiliateDiscountStrategy.kt
package com.example.retailstore.strategy

import com.example.retailstore.model.Bill
import org.springframework.stereotype.Component


@Component
class AffiliateDiscountStrategy : DiscountStrategy {
    override fun calculateNetPayableAmount(bill: Bill): Double {
        val totalAmount = bill.totalAmount
        var netPayableAmount = totalAmount
        netPayableAmount -= totalAmount * AFFILIATE_DISCOUNT

        val discountAmount = (totalAmount / 100).toInt() * DiscountStrategy.DISCOUNT_PER_100
        netPayableAmount -= discountAmount

        return netPayableAmount
    }

    companion object {
        private const val AFFILIATE_DISCOUNT = 0.1
    }
}
